File: solutions/go/game-of-life/1/game_of_life.go
package gameoflife

func at(matrix [][]int, x, y int) int {
	if y < 0 || y >= len(matrix) || x < 0 || x >= len(matrix[0]) {
		return 0
	}
	return matrix[y][x]
}

func neighbors(matrix [][]int, x, y int) int {
	count := 0
	for dx := -1; dx <= 1; dx++ {
		for dy := -1; dy <= 1; dy++ {
			if dx != 0 || dy != 0 {
				count += at(matrix, x+dx, y+dy)
			}
		}
	}
	return count
}

func Tick(matrix [][]int) [][]int {
	next := make([][]int, len(matrix))
	for i := range matrix {
		next[i] = make([]int, len(matrix[i]))
		copy(next[i], matrix[i])
	}

	for y := 0; y < len(matrix); y++ {
		for x := 0; x < len(matrix[0]); x++ {
			neighborCount := neighbors(matrix, x, y)
			current := at(matrix, x, y)

			if current == 1 && (neighborCount <= 1 || neighborCount >= 4) {
				next[y][x] = 0
			}
			if current == 0 && neighborCount == 3 {
				next[y][x] = 1
			}
		}
	}
	return next
}

File: solutions/python/game-of-life/1/game_of_life.py
"""Conway's Game of Life."""


def tick(matrix):
    """Return the next generation of the given Game of Life board."""
    rows = len(matrix)
    cols = len(matrix[0]) if rows else 0

    def neighbors(r, c):
        count = 0
        for dr in (-1, 0, 1):
            for dc in (-1, 0, 1):
                if dr == 0 and dc == 0:
                    continue
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols:
                    count += matrix[nr][nc]
        return count

    return [
        [
            int(
                neighbors(r, c) == 3
                or (matrix[r][c] == 1 and neighbors(r, c) == 2)
            )
            for c in range(cols)
        ]
        for r in range(rows)
    ]